File: 03_patterns/swift/mvvm_clean_arc/Clean/Presentation/ViewModels/CleanTaskListViewModel.swift
import Combine
import Foundation

@MainActor
final class CleanTaskListViewModel: ObservableObject {
    @Published private(set) var tasks: [TaskItem] = []

    private let getTasksUseCase: GetTasksUseCase
    private let createTaskUseCase: CreateTaskUseCase
    private let deleteTaskUseCase: DeleteTaskUseCase
    private let taskItemMapper: TaskItemMapper

    init(
        getTasksUseCase: GetTasksUseCase,
        createTaskUseCase: CreateTaskUseCase,
        deleteTaskUseCase: DeleteTaskUseCase,
        taskItemMapper: TaskItemMapper = TaskItemMapper()
    ) {
        self.getTasksUseCase = getTasksUseCase
        self.createTaskUseCase = createTaskUseCase
        self.deleteTaskUseCase = deleteTaskUseCase
        self.taskItemMapper = taskItemMapper
        loadTasks()
    }

    func loadTasks() {
        tasks = taskItemMapper.map(getTasksUseCase.execute())
    }

    func createTask(title: String) {
        let trimmedTitle = title.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedTitle.isEmpty else { return }

        tasks = taskItemMapper.map(createTaskUseCase.execute(title: trimmedTitle))
    }

    func deleteTask(at offsets: IndexSet) {
        let selectedTaskIDs = offsets.map { tasks[$0].id }

        for taskID in selectedTaskIDs {
            tasks = taskItemMapper.map(deleteTaskUseCase.execute(id: taskID))
        }
    }
}

File: 03_patterns/flutter/lib/clean/presentation/view_models/clean_task_list_view_model.dart
import 'package:flutter/foundation.dart';

import '../../../shared/models/task_item.dart';
import '../../domain/use_cases/create_task_use_case.dart';
import '../../domain/use_cases/delete_task_use_case.dart';
import '../../domain/use_cases/get_tasks_use_case.dart';
import '../mappers/task_item_mapper.dart';

class CleanTaskListViewModel extends ChangeNotifier {
  CleanTaskListViewModel({
    required GetTasksUseCase getTasksUseCase,
    required CreateTaskUseCase createTaskUseCase,
    required DeleteTaskUseCase deleteTaskUseCase,
    TaskItemMapper taskItemMapper = const TaskItemMapper(),
  }) : _getTasksUseCase = getTasksUseCase,
       _createTaskUseCase = createTaskUseCase,
       _deleteTaskUseCase = deleteTaskUseCase,
       _taskItemMapper = taskItemMapper {
    loadTasks();
  }

  final GetTasksUseCase _getTasksUseCase;
  final CreateTaskUseCase _createTaskUseCase;
  final DeleteTaskUseCase _deleteTaskUseCase;
  final TaskItemMapper _taskItemMapper;

  List<TaskItem> _tasks = const [];
  List<TaskItem> get tasks => List.unmodifiable(_tasks);

  void loadTasks() {
    _tasks = _taskItemMapper.mapList(_getTasksUseCase.execute());
    notifyListeners();
  }

  void createTask({required String title}) {
    final trimmedTitle = title.trim();
    if (trimmedTitle.isEmpty) {
      return;
    }

    _tasks = _taskItemMapper.mapList(
      _createTaskUseCase.execute(title: trimmedTitle),
    );
    notifyListeners();
  }

  void deleteTask({required int index}) {
    if (index < 0 || index >= _tasks.length) {
      return;
    }

    final taskId = _tasks[index].id;
    _tasks = _taskItemMapper.mapList(
      _deleteTaskUseCase.execute(id: taskId),
    );
    notifyListeners();
  }
}
